synchronicity/synchronizer.py

```diff
@@ -620,16 +625,22 @@ def _wrap(
         # It wraps the object, and caches the wrapped object
 
         # Get the list of existing interfaces
-        if self._wrapped_attr not in obj.__dict__:
-            if isinstance(obj.__dict__, dict):
-                # This works for instances
-                obj.__dict__.setdefault(self._wrapped_attr, {})
-            else:
-                # This works for classes & functions
+        if hasattr(obj, "__dict__"):
+            if self._wrapped_attr not in obj.__dict__:
+                if isinstance(obj.__dict__, dict):
+                    # This works for instances
+                    obj.__dict__.setdefault(self._wrapped_attr, {})
+                else:
+                    # This works for classes & functions
+                    setattr(obj, self._wrapped_attr, {})
+            interfaces = obj.__dict__[self._wrapped_attr]
+        else:
+            # e.g., TypeVar in Python>=3.12
+            if not hasattr(obj, self._wrapped_attr):
```

Comment:
It's not obvious to me why we are preferring to update `obj.__dict__` rather than use `setattr` everywhere. I assume there is a good reason so I have preserved the existing code, making this somewhat verbose. If there's not a good reason, we can simplify.

Reply:
As you pointed out in a separate thread, when the wrapped classes define `__setattr__` hooks, things could get messy otherwise.

We do have a getattr_test.py unit test that actually fails when I change the code to use setattr for object instances

Reply:
Makes sense; the resulting code is cumbersome but not unbearably so imo.